Redirect app updates instead of syncs/submissions

kaapstorm opened the pull request:
## Product Description

This change allows domains that have been migrated to third-party-managed hosts to redirect app updates to their new host.

This avoids problems when deployed CommCare apps are not using a DNS alias prior to the migration.

## Technical Summary

Based on discussions with the Mobile team, we are adopting a slightly different approach from that taken in PR https://github.com/dimagi/commcare-hq/pull/34461

Instead of redirecting restores/syncs and form submissions, we are redirecting app updates instead.

Jira: [SC-3628](https://dimagi.atlassian.net/browse/SC-3628)

Documentation: [Redirect on Data Migration](https://docs.google.com/document/d/1mQxRmboW3CuOkmklUjf5jypgJcv9VX_e26TEiFh1CyI/edit)

FYI @avazirna 

## Feature Flag

Data Migration

## Safety Assurance

### Safety story

HQ behavior verified by unit tests. Tested in collaboration with Mobile team.

### Automated test coverage

Tests included

### QA Plan

QA ticket: [QA-6406](https://dimagi.atlassian.net/browse/QA-6406)

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SC-3628]: https://dimagi.atlassian.net/browse/SC-3628?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[QA-6406]: https://dimagi.atlassian.net/browse/QA-6406?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kaapstorm committed:
Redirect app updates instead of syncs/submissions

millerdev commented on corehq/apps/domain/decorators.py:
```diff
@@ -671,13 +670,6 @@ def _inner(request, *args, **kwargs):
     def wrapped_view(request, domain, *args, **kwargs):
-        domain_obj = Domain.get_by_name(domain)
-        if domain_obj.redirect_url:
-            # IMPORTANT!
-            #     We assume that the domain name is the same on both
-            #     environments.
```
Should this comment be preserved in the new decorator?

kaapstorm replied:
We are making more assumptions this time:
* The domain name is the same on both
* The app exists on both

Both of these will be true for a migrated environment. So instead of making the comment longer, I figured I'd take it out, and cover the purpose of this feature in the documentation.

kaapstorm replied:
[Documentation](https://docs.google.com/document/d/1mQxRmboW3CuOkmklUjf5jypgJcv9VX_e26TEiFh1CyI/edit) updated.

millerdev commented on corehq/apps/domain/tests/test_redirect_url.py:
```diff
@@ -112,65 +110,52 @@ def _call_redirect_url(*args, **kwargs):
             )
 
-    @flag_enabled('DATA_MIGRATION')
-    def test_service_unavailable_response(self):
```
Is this test no longer relevant because requests like the one it was testing are now redirected?

kaapstorm replied:
Good question. https://github.com/dimagi/commcare-hq/pull/34505/commits/88b8b676dcb66d4403f937fbe652e52cd8ddbbd4

mkangia commented:
> Instead of redirecting restores/syncs and form submissions, we are redirecting app updates instead.

Cool, I did wonder in that last PR why we are not addressing app updates instead. Just realized that I never put that out on the PR as a query.

kaapstorm committed:
Verify behavior when Data Migration enabled

kaapstorm committed:
Wrap check_redirect outside safe_cached_download

`safe_cached_download` strips `username` from request.GET, and we need
it, so check the redirect first.

kaapstorm committed:
Update command help

Charl1996 commented:
@kaapstorm Please correct me if I'm wrong, but I'll quickly write out what my understanding is of the current workflow from the user's perspective:
User A uses a commcare app which is still pointing to `commcarehq.org`, but the project has successfully migrated their data to an on-prem site. User A now wants to sync, but commcarehq.org will deny the request since the project has the DATA_MIGRATION flag enabled. What User A now needs to do is simply update their app, since the project presumably has a new updated app available on their on-prem site, which the user will be able to access through this redirect process.  

Is this a correct summary?

kaapstorm commented:
@Charl1996 Yes, spot on. As long as the new environment has an updated app, that's exactly what will happen. CommCare Mobile will upgrade the app, and the updated app will use the correct URLs to submit forms / sync. (The user will not be able to sync until an updated app is available.)

Charl1996 reviewed (approved):
Looks good.
